map.go

```diff
@@ -142,32 +142,29 @@ func (m *Map) Load(key string) (value interface{}, ok bool) {
 			if key == derefKey(kp) {
 				if uintptr(vp) == uintptr(atomic.LoadPointer(&b.values[i])) {
 					// Atomic snapshot succeeded.
-					return derefValue(vp), true
+					return derefValue[V](vp), true
 				}
 				// Concurrent update/remove. Go for another spin.
 				goto atomic_snapshot
 			}
 		}
 	}
-	return nil, false
+	return value, false
 }
 
 // Store sets the value for a key.
-func (m *Map) Store(key string, value interface{}) {
+func (m *MapOf[V]) Store(key string, value V) {
 	m.doStore(key, value, false)
 }
 
 // LoadOrStore returns the existing value for the key if present.
 // Otherwise, it stores and returns the given value.
 // The loaded result is true if the value was loaded, false if stored.
-func (m *Map) LoadOrStore(key string, value interface{}) (actual interface{}, loaded bool) {
+func (m *MapOf[V]) LoadOrStore(key string, value V) (actual V, loaded bool) {
 	return m.doStore(key, value, true)
 }
 
-func (m *Map) doStore(key string, value interface{}, loadIfExists bool) (actual interface{}, loaded bool) {
-	if value == nil {
-		value = nilVal
-	}
+func (m *MapOf[V]) doStore(key string, value V, loadIfExists bool) (actual V, loaded bool) {
```

Comment:
Using plain reference type (`*V`) for values instead of `*interface{}` -> `*V` will break the atomic snapshot logic used by the hash map readers. The thing is that the set of live `*interface{}` references forms a unique set. This property is used by readers that read and later compare value pointers when reading the key/value pair atomically. The atomic snapshot logic is basically what happens in this loop:
https://github.com/puzpuzpuz/xsync/blob/efd8a81aa9261ce9d19dc9c02a9a87e8f34d8e93/map.go#L137-L151

#8 describes the problem in more detail and a way to approach it.

Reply:
Thank you for explanation, I still don't understand the mechanics here. I thought the set of live `*V` would also form a unique set (similarly to `*interface{}` as each pointer to `V` would be unique), and will have same size (8 per each pointer).

Would `uintptr(vp) == uintptr(atomic.LoadPointer(&b.values[i]))` be a false positive with `*V`?

Reply:
> I thought the set of live `*V` would also form a unique set (similarly to `*interface{}` as each pointer to `V` would be unique), and will have same size (8 per each pointer).

Same size - no doubt. As for the unique set, I need to refresh my memory on that. You may be correct, but I'd like to double check.

Reply:
After some experimenting, I think that it's ok to get rid of the empty interface and use arbitrary types for values. My concern was around pointer types and built-in reference types (map, slice, channel). But it looks like an intermediate struct is being allocated on the heap for such values by `unsafe.Pointer(&value)` (BTW this may explain why you saw no difference in the benchmarks).

Also, the spec doesn't help a lot in this respect since it only [mentions](https://go.dev/ref/spec#Calls) the following:
> After they are evaluated, the parameters of the call are *passed by value* to the function and the called function begins execution.

Here is the program I was experimenting with:
```go
package main

import (
	"fmt"
	"unsafe"
)

type T struct {
	x, y int
}

var tPtr unsafe.Pointer

func main() {
	t1 := new(T)
	testPtr(t1)
	testPtr(t1)

	t2 := new(T)
	testPtr(t2)
}

//go:noinline
func testPtr(t *T) {
	tPtr = unsafe.Pointer(&t)
	fmt.Printf("Pointer value: %d\n", uintptr(tPtr))
}
```

Interestingly, all of the three pointer addresses are different.

Anyhow, I don't see any problems with supporting generic in the map. This would be a great addition.

Reply:
Great, then I will proceed with proper decoupling of `MapOf` from `Map` based on these changes.

Reply:
That would be awesome. Thanks 👍